protocol/src/lib.rs

```diff
@@ -941,7 +976,11 @@ impl<'a> Handshake<'a> {
 #[cfg(feature = "std")]
 #[derive(Debug)]
 pub enum ProtocolError {
+    /// Wrap all IO errors.
     Io(std::io::Error),
+    /// Remote likely only supports V1 protocol.
+    IncompatibleProtocol(std::io::Error),
```

Comment:
I'm still not fully convinced of this error variant. I think the caller should be responsible for deciding if they want to retry with version one.

Reply:
Hm yea, maybe there is a less heavy-handed approach to this? I'll dump my thoughts here for posterity, but open to just removing it.

This is attempting to codify this [blurb in BIP324](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#signaling-v2-support)

> If met with immediate disconnection when establishing a v2 connection, clients implementing this proposal are encouraged to retry connecting using the v1 protocol.

Ideally this is somehow captured in the code so a caller doesn't have to know this by reading the BIP. The next best thing would be to document it somewhere in the library, but I think it is worth getting it in the code in the case to ensure a caller is aware.

Not sure if there are better patterns for this, but I went with a separate variant to make it easy for a caller to match on. Maybe this could instead be a field on the `Io` variant?

Options:

1. New variant
2. New field on `Io` variant
3. Just documentation on the `Io` variant